### ISSUE TITLE
helm command with --kube-apiserver --kube-token

### ISSUE THREAD

**Issue by chancejin** [helm/helm#8487]
Helm version : 3.2.0
Target K8s server version : 1.17
Using Openshift.

I tried to use helm command in my local.

helm list --kubeconfig {kubeconfigFilePath} --kube-context {contextName}
-> It works fine.

helm list --kube-apiserver https://{API server IP}:{API server Port} --kube-token {Kubetoken}
-> It shows error like below

Error: Kubernetes cluster unreachable: Get https://10.251.205.11:6443/version?timeout=32s: x509: certificate signed by unknown authority"

Is there necessary data do I have to add to use "--kube-apiserver" without "--kubeconfig"?

**Comment by chancejin** [comment]
@bacongobbler 
https://github.com/helm/helm/pull/7230

**Comment by bacongobbler** [comment]
cc'ing @waveywaves as he was working on this functionality and might have a better idea what's causing this.

**Comment by donggangcj** [comment]
I reproduced this issue.This is due to a certificate validation failure. `kubectl` can pass  the flag ` --insecure-skip-tls-verify`to skip verification or pass the  certificate file. The related flag of `kubectl` is here:

```
➜  .kube kubectl options
The following options can be passed to any command:
     ...
      --certificate-authority='': Path to a cert file for the certificate authority
      --client-certificate='': Path to a client certificate file for TLS
      --client-key='': Path to a client key file for TLS
      --insecure-skip-tls-verify=false: If true, the server's certificate will not be checked for validity. This will
make your HTTPS connections insecure 
      -s, --server='': The address and port of the Kubernetes API server
      --token='': Bearer token for authentication to the API server
```
If it is possible，I can fix it.

**Comment by chancejin** [comment]
I cannot use insecure connection.
Is there any options for passing certificate file in helm ?

**Comment by donggangcj** [comment]
@chancejin  There seems to be no way to passing certificate for now. What situations have you encountered where you must
paas token instead of use `kubeconfig file`?

**Comment by chancejin** [comment]
@donggangcj  Thank you for answering first.

I want to use REST API to getting information.
But I cannot find the way to get kubeconfig file information through REST API.

**Comment by donggangcj** [comment]
Maybe you can construct kubeconfig manually. 🙂

**Comment by emreyokur** [comment]
+1 for the feature request. 

In my opinion, both, skipping the certificate check and providing a global flag to pass certificate needs to be implemented to remove the dependence of kubeconfig file of helm to authenticate to kubernetes.

**Comment by AdarShaked** [comment]
+1 for this feature. It is going to be merged soon?

**Comment by donggangcj** [comment]
> +1 for this feature. It is going to be merged soon?
> @donggangcj you said that seems to be no way to passing certificate for now for cluster.

@AdarShaked  Yes, certificate delivery is not currently supported and  I will create a new PR to support this feature.

> so what are flags "--ca-file,--cert-file,--key-file" in install command used for?downloading the chart only?
> @bridgetkromhout maybe you have answer for that?

Yeap! These flags are used for the chart repo.

**Comment by AdarShaked** [comment]
thanks, 
@donggangcj 
when https://github.com/helm/helm/pull/8622 will be merged?

**Comment by donggangcj** [comment]
@AdarShaked I'm not a maintainer and #8622 need a Review. 
cc@hickeyma  @bacongobbler

**Comment by AdarShaked** [comment]
@hickeyma @bacongobbler 
when this pr  planed for review?

**Comment by AdarShaked** [comment]
> @AdarShaked Yes, certificate delivery for kubernetes APIServer is not currently supported and I will create a new PR to support this feature.

Hi , you opened PR for the certificates fuature?

**Comment by donggangcj** [comment]
@AdarShaked  not yet.

**Comment by bacongobbler** [comment]
As a note, #8622 does not close out this issue. As the original issue suggests, flags for `--certificate-authority`, `--client-certificate`, and `--client-key` need to be implemented as well. #8622 only introduces `--insecure-skip-tls-verify`.

**Comment by bacongobbler** [comment]
Updated the title.

**Comment by donggangcj** [comment]
Yeap,#8622  just offering a solution to fix this issue and the problem of certification with Kubernetes is not fully resolved. 
I will open a feature PR to improve this function as @bacongobbler suggested.

**Comment by jmcelwain** [comment]
Can #8622 be reviewed and merged without solutions for the other certificate related features?

**Comment by unkis** [comment]
any updates on this?

**Comment by sysadm1989** [comment]
Why aren't we implementing arguments to support the certificate?

**Comment by rakyll** [comment]
+1 to this feature.

**Comment by AdarShaked** [comment]
Hi, any updates for --kube-insecure-skip-tls-verify ?

**Comment by philomory** [comment]
This is not stale.

**Comment by ozbillwang** [comment]
With current version, I do see the option

```
$ helm install --help |grep insecure-skip
      --insecure-skip-tls-verify     skip tls certificate checks for the chart download
```

But seems it doesn't work as expect.

```
$ TOKEN=$(kubectl get secret $(kubectl get secret | grep cicd-token | awk '{print $1}') -o jsonpath='{.data.token}' | base64 --decode)
$ KUBE_APISERVER="https://xxx.example.com"

$ kubectl --insecure-skip-tls-verify --kubeconfig="/dev/null" --server=${KUBE_APISERVER}  --token=$TOKEN  get pods

NAME                            READY   STATUS             RESTARTS   AGE
hello                           1/1     Running            1          2d

# So I can see the pods 

# But when I convert the similar options to helm chart command, it failed.

$ helm upgrade --install --insecure-skip-tls-verify \
  --kubeconfig="/dev/null" \
  --kube-apiserver=${KUBE_APISERVER} \
  --kube-token=${TOKEN} \
  -n best \
  my-release .

WARNING: Kubernetes configuration file is group-readable. This is insecure. Location: /dev/null
WARNING: Kubernetes configuration file is world-readable. This is insecure. Location: /dev/null
Error: Kubernetes cluster unreachable: Get "https://xxx.example.com/version?timeout=32s": x509: certificate signed by unknown authority

```

So does the option `--insecure-skip-tls-verify` work?

ref: https://www.digitalocean.com/community/tutorials/how-to-automate-deployments-to-digitalocean-kubernetes-with-circleci

**Comment by 0xERR0R** [comment]
> --insecure-skip-tls-verify     skip tls certificate checks for the chart download

`--insecure-skip-tls-verify     skip tls certificate checks for the chart download` -> this is only for charts download, this is not used for kubernetes API

**Comment by charlspjohn** [comment]
+1 to this feature

**Comment by anatoly-kryzhanovsky** [comment]
+1 to this feature

**Comment by gSerP1983** [comment]
+1

**Comment by yuanxc0221** [comment]
+10086

**Comment by bacongobbler** [comment]
In the interest of preventing spam to users subscribing to this thread, please add a :+1: to the OP instead of commenting "+1". That way we can keep the conversation relevant to the discussion at hand. I've gone ahead and deleted those comments. Thanks!

**Comment by heumsi** [comment]
This is not stale.

Any new update?

**Comment by heumsi** [comment]
This is not stale

**Comment by heumsi** [comment]
This is not stale.
